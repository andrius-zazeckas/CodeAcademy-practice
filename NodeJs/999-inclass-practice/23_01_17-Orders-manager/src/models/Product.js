import Joi from "joi";

const productSchema = Joi.object({
  id: Joi.number(),
  name: Joi.string().trim().required(),
  price: Joi.number().required(),
  isAvailable: Joi.boolean().required(),
  imageUrl: Joi.string().trim().required(),
});

export class Product {
  id;
  name;
  price;
  isAvailable;
  imageUrl;

  constructor(name, price, isAvailable, imageUrl, id) {
    const newProductData = { id, name, price, isAvailable, imageUrl };

    const validationResult = productSchema.validate(newProductData);

    if (validationResult.error) {
      throw Error(validationResult.error);
    }

    this.id = id;
    this.name = name;
    this.price = price;
    this.isAvailable = isAvailable;
    this.imageUrl = imageUrl;
  }
}
