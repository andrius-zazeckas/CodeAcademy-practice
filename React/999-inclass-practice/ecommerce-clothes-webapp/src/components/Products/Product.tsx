import { ImageListItem, ImageListItemBar } from "@mui/material";
import { type FC, useContext } from "react";
import { ProductsContext } from "../ProductsContext";
import type { TProductProps } from "./types";

// export const Product = ({ product }: TProductProps) => {
export const Product: FC<TProductProps> = ({ product }) => {
  const { dispatch } = useContext(ProductsContext);

  return (
    <ImageListItem>
      <img
        src={`${product.image}?w=248&fit=crop&auto=format`}
        srcSet={`${product.image}?w=248&h&fit=crop&auto=format&dpr=2 2x`}
        loading="lazy"
      />

      <ImageListItemBar
        title={product.title}
        subtitle={<span>PRICE: {product.price}</span>}
        position="top"
      />

      <button
        onClick={() =>
          dispatch({ type: "addProduct", payload: { productId: product.id } })
        }
      >
        Add to cart
      </button>

      <button
        onClick={() =>
          dispatch({
            type: "removeProduct",
            payload: { productId: product.id },
          })
        }
      >
        Remove from cart
      </button>
    </ImageListItem>
  );
};
