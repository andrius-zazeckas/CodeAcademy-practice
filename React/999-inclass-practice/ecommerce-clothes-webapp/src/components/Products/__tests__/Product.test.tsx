import { render, screen } from "@testing-library/react";
import { Product } from "../Product";
import renderer from "react-test-renderer";

describe("Product", () => {
  const product = {
    id: 1,
    title: "Test",
    price: 2.99,
    description: "For testing purposes",
    category: "Testing category",
    image: "testImage.jpg",
    rating: {
      rate: 4,
      count: 2,
    },
  };

  it("should render Product", () => {
    render(<Product product={product} />);
    expect(screen.getByLabelText("product")).toBeVisible();

    const tree = renderer.create(<Product product={product} />);

    expect(tree).toMatchSnapshot();
  });

  it("should show product information", () => {
    render(<Product product={product} />);

    const productTitle = screen.getByLabelText("product title");
    const productImage = screen.getByRole("img");
    const productPrice = screen.getByLabelText("product price");

    expect(productTitle).toBeVisible();
    expect(productImage).toBeVisible();
    expect(productImage).toHaveAttribute("src", product.image);

    expect(productPrice).toBeVisible();

    const tree = renderer.create(<Product product={product} />);

    expect(tree).toMatchSnapshot();
  });
});
