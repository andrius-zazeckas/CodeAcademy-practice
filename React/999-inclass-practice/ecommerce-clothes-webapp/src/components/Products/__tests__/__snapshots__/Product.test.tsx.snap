// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should render Product 1`] = `
<div
  aria-label="product"
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-3 css-vrw0vt-MuiGrid-root"
>
  <div
    aria-label="product image"
    className="MuiBox-root css-civhbh"
  >
    <img
      alt="Test"
      src="testImage.jpg"
    />
  </div>
  <div
    className="MuiBox-root css-1imvbg5"
  >
    <p
      aria-label="product title"
      className="MuiTypography-root MuiTypography-body1 css-zwkcuf-MuiTypography-root"
    >
      Test
    </p>
    <p
      aria-label="product price"
      className="MuiTypography-root MuiTypography-body1 css-mfsdrg-MuiTypography-root"
    >
      PRICE: 
      2.99
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      For testing purposes
    </p>
  </div>
  <div
    className="MuiBox-root css-1xcgjaw"
  >
    <button
      aria-label="add product"
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1klp1e9-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      +
    </button>
  </div>
</div>
`;

exports[`Product should show product information 1`] = `
<div
  aria-label="product"
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-3 css-vrw0vt-MuiGrid-root"
>
  <div
    aria-label="product image"
    className="MuiBox-root css-civhbh"
  >
    <img
      alt="Test"
      src="testImage.jpg"
    />
  </div>
  <div
    className="MuiBox-root css-1imvbg5"
  >
    <p
      aria-label="product title"
      className="MuiTypography-root MuiTypography-body1 css-zwkcuf-MuiTypography-root"
    >
      Test
    </p>
    <p
      aria-label="product price"
      className="MuiTypography-root MuiTypography-body1 css-mfsdrg-MuiTypography-root"
    >
      PRICE: 
      2.99
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      For testing purposes
    </p>
  </div>
  <div
    className="MuiBox-root css-1xcgjaw"
  >
    <button
      aria-label="add product"
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1klp1e9-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      +
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
